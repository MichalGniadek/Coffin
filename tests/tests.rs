use coffin2::{debug_print, lexer, name_resolution, parser, type_resolution};
use insta::{assert_snapshot, glob};
use std::{fs, path::Path};

fn test_util(path: &Path, spans: bool, variables: bool, types: bool) -> String {
    let code = fs::read_to_string(path).unwrap();
    let lexer = lexer::lex(&code);
    let mut ast = parser::parse(lexer);

    let variables = if variables | types {
        Some(name_resolution::visit(&mut ast))
    } else {
        None
    };

    let types = if types {
        Some(type_resolution::visit(
            &mut ast,
            variables.as_ref().unwrap(),
        ))
    } else {
        None
    };

    // let _module = spirv_generation::visit(&mut ast, &vars, &type_table);

    debug_print::visit(&ast, true, spans, variables.as_ref(), types.as_ref())
}

#[test]
fn parser() {
    glob!(r"insta_parser\*.coff", |path| {
        assert_snapshot!(test_util(path, false, false, false));
    });
}

#[test]
fn parser_spans() {
    glob!(r"insta_parser_spans\*.coff", |path| {
        assert_snapshot!(test_util(path, true, false, false));
    });
}

#[test]
fn variables() {
    glob!(r"insta_variables\*.coff", |path| {
        assert_snapshot!(test_util(path, false, true, false));
    });
}

#[test]
fn types() {
    glob!(r"insta_types\*.coff", |path| {
        assert_snapshot!(test_util(path, false, false, true));
    });
}
