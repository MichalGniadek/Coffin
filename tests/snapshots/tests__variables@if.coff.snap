---
source: tests/tests.rs
expression: "test_util(path, false, true, false)"

---
fun test@0 ()  {
	(if $a@1 {
		$a@2
	} else {
		$a@3
	})
}

===
Undeclared variable. [20..21]
Undeclared variable. [23..24]
Undeclared variable. [32..33]
